### DISCUSSION TITLE
Apple game porting toolkit & shader porting

### DISCUSSION THREAD

**Discussion by magnuslundstedt** [KhronosGroup/MoltenVK#1937] [General]
Can there be any new changes in Metal API for MacOS Sonoma that helps VulkanVK?

Looking for clues in the guide for porting shaders there seem to be new changes in Metal API (as ported shaders require Sonoma): [https://developer.apple.com/metal/shader-converter/](https://developer.apple.com/metal/shader-converter/) 

One thing that stood out to me was support for bindless resources. I’m no Metal or Vulkan expert, but curios to know if anyone see changes coming to Metal, and/or think upcoming changes may help VulkanVK forward?

Edit: Looking at the new Metal shader documentation, they are releasing Metal 3.1 in macOS Sonoma: 
[https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf)

**Comment by IsaacMarovitz** [comment]
The kit and converter are specifically for DirectX not Vulkan

**Reply by billhollings** [reply]
> The kit and converter are specifically for DirectX not Vulkan

However, MoltenVK does support loading MSL binaries through `VkShaderModule` (although not as straightforward as it could be...there is some challenge in aligning resource index numbers). So this might have values in two ways:

- Devs using the Metal shader converter to port DXIL to MSL binaries, which the app could then load into MoltenVK.
- Enhancing MoltenVK to accept DXIL input via `VkShaderModule`, and then convert them internally into MSL binaries.

The goal of both of these would be to bypass the SPIRV->MSL->MSL binary conversions and compilations.

**Reply by airsquared** [reply]
Based on [this video](https://developer.apple.com/videos/play/wwdc2023/10124) they do seem to provide at least a little more support for geometry shaders, or a way to make it easier to implement? At 7:49 they talk about how all the unsupported shaders like geometry shaders are mapped to mesh shaders and how the new `objectLinkedFunctions` and `meshLinkedFunctions` supports this.

**Reply by UltraHDR** [reply]
It'll be interesting to see if MoltenVK can support converting SPIR-V -> DXIL -> Metal IR (https://gitlab.freedesktop.org/mesa/mesa/-/tree/main/src/microsoft/spirv_to_dxil & https://developer.apple.com/metal/shader-converter). Yes it'll take double the time but it might yield noticeable performance/compatability improvements.

**Reply by billhollings** [reply]
> It'll be interesting to see if MoltenVK can support converting SPIR-V -> DXIL -> Metal IR 

Yeah. I realized that might be great path if we can get it working. I've started some basic research on it.
